app/grandchallenge/emails/tasks.py

```diff
@@ -0,0 +1,30 @@
+from celery import shared_task
+from django.contrib.sites.models import Site
+from django.core.mail import send_mass_mail
+from django.core.paginator import Paginator
+
+from config import settings
+from grandchallenge.subdomains.utils import reverse
+
+
+@shared_task
+def send_bulk_email(users, subject, body):

```

Comment:
For designing Celery tasks you should be aware that there are some restrictions. The main one is that they're idempotent, so that if they're called twice with the same arguments the work only happens once, but you know that already. Once you are sure that the task is idempotent you can use one of the acks-late queues, which makes the task more robust to failure (it is best to avoid the standard queue).

Another, is that the arguments to the task are passed through a JSON serializer, so sending Python objects (in this case, users) is really tricky, you need to stick to basic types (numbers, strings, etc) and then within your task fetch all the information that is needed. Due to the SQS broker we use, there is also a limit on how large the message containing all of the arguments can be (2MB), so it is best to avoid sending lists of things, instead, send enough information so that the list can be generated in the task instead (for instance, using an Enum choice).

Finally, for all Celery tasks, I prefer to use keyword-only arguments as it is easier to maintain (the same goes for almost all functions).

Reply:
It's only really about timing, if you think that it's going to take more than 300s then use the large queue.